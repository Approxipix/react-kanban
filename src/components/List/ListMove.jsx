import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { moveList, moveListToAnotherBoard } from '../../redux/boardReducer/actions/actions';
import { SubmitButton } from '../BaseComponent';
import Select from '../Select'
import styled from 'styled-components'

const Container = styled.div`
  margin-bottom: 1rem;
`;

class ListDelete extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newBoardId: props.boardId || null,
      newListPosition: 1,
    }
  }

  moveList =() => {
    const { boards, boardId, listId } = this.props;
    const { newBoardId, newListPosition } = this.state;
    const currentListPosition = boards[boardId].lists.indexOf(listId);
    if (newBoardId === boardId && currentListPosition === newListPosition) {
      this.props.toggleOpened();
      return null;
    }
    if (boardId === newBoardId) {
      this.props.actions.moveList({
        boardId: boardId,
        sourceIndex: boards[boardId].lists.indexOf(listId),
        destinationIndex: newListPosition - 1,
      })
    } else {
      this.props.actions.moveListToAnotherBoard({
        boardId: boardId,
        newBoardId: newBoardId,
        listId: listId,
        newListPosition: newListPosition - 1,
      })
    }
  };

  setNewBoardId = (id) => {
    const { boards } = this.props;
    const { newListPosition } = this.state;
    const listLength = boards[id].lists.length;
    if (listLength < newListPosition) {
      this.setState({
        newListPosition: listLength || 1,
      })
    }
    this.setState({
      newBoardId: id,
    });
  };

  boardIdSelector = () => {
    const { newBoardId } = this.state;
    const { boards, boardId } = this.props;
    const boardsOptions = Object.keys(boards).map((id) => (
      {
        title: boardId === id
          ? `${boards[id].title} (current)`
          : boards[id].title,
        value: id,
      }
    ));
    return (
      <Select
        label="Board"
        placeholder="Select board"
        value={boards[newBoardId].title}
        options={boardsOptions}
        onChange={(id) => this.setNewBoardId(id)}
      />
    );
  };

  listPositionSelector = () => {
    const { newBoardId, newListPosition } = this.state;
    const { boards, boardId, listId } = this.props;
    const currentListPosition = boards[boardId].lists.indexOf(listId);
    let listLength = boards[newBoardId].lists.length;
    if (boardId !== newBoardId) listLength = listLength + 1;
    let positionOptions = [...Array(listLength).keys()];
    positionOptions = positionOptions.map(position => (
      {
        title: currentListPosition === position && boardId === newBoardId
          ? `${++position} (current)`
          : ++position,
        value: position,
      }
    ));
    return (
      <Select
        label="Position"
        placeholder="Select position"
        value={newListPosition}
        options={positionOptions}
        onChange={(position) => this.setState({
          newListPosition: position,
        })}
      />
    )
  };

  render() {
    return (
      <>
        <Container>
          {this.boardIdSelector()}
          {this.listPositionSelector()}
        </Container>
        <SubmitButton onClick={() => this.moveList()}>
          Move
        </SubmitButton>
      </>
    );
  }
}

function mapStateToProps(state) {
  return {
    boards: state.boardReducer,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      moveList: moveList,
      moveListToAnotherBoard: moveListToAnotherBoard,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ListDelete);
