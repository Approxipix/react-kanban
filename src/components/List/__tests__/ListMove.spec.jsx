import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListMove from '../ListMove';

const mockStore = configureStore([]);

describe('<ListMove>', () => {
  let store;
  let props;
  let component;
  let mockToggleOpened= jest.fn();
  let mockMoveList = jest.fn();
  let mockMoveListToAnotherBoard = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    store = mockStore({
      boardReducer: {
        Board1ID: {
          _boardId: 'Board1ID',
          title: 'Board 1',
          lists: ['List11', 'List12', 'List13'],
        },
        Board2ID: {
          _boardId: 'Board2ID',
          title: 'Board 2',
          lists: ['List21', 'List22', 'List23'],
        }
      }
    });
    store.dispatch = mockMoveList;
    props = {
      boardId: 'Board1ID',
      listId: 'List11',
      toggleOpened: mockToggleOpened
    };
    const state = {
      newBoardId: 'Board1ID',
      newListPosition: 0,
    };

    const wrapper = shallow(<ListMove store={store} {...props} />);
    component = wrapper.find('ListMove').dive();
    component.setState({ ...state });
  });

  it('should render with given state from Redux store and correct props', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if setNewBoardId sets newBoardId correctly', () => {
    component.instance().setNewBoardId('Board2ID');

    expect(component.instance().state.newBoardId).toEqual('Board2ID');
  });

  it('checks if setNewBoardId sets newListPosition correctly', () => {
    component.setState({ newListPosition: 5 });
    component.instance().setNewBoardId('Board2ID');

    expect(component.instance().state.newListPosition).toEqual(3);
  });

  it('should call toggleOpened if boardId/newBoardId and currentListPosition/newListPosition are the same', () => {
    component.find('SubmitButton').simulate('click');

    expect(mockToggleOpened).toHaveBeenCalledTimes(1);
  });

  describe('dispatch action', () => {
    it("should dispatch an action if boardId/newBoardId are the same", () => {
      const expectedAction = {
        type: 'MOVE_LIST',
        payload: {
          boardId: 'Board1ID',
          sourceIndex: 0,
          destinationIndex: 1,
        }
      };

      component.setState({ newListPosition: 2 });
      component.find('SubmitButton').simulate('click');

      expect(mockMoveList).toHaveBeenCalledTimes(1);
      expect(mockMoveList).toHaveBeenCalledWith(expectedAction);
    });

    it("should dispatch an action if boardId/newBoardId are different", () => {
      const expectedAction = {
        type: 'MOVE_LIST_TO_ANOTHER_BOARD',
        payload: {
          boardId: 'Board1ID',
          newBoardId: 'Board2ID',
          listId: 'List11',
          newListPosition: 0,
        }
      };
      store.dispatch = mockMoveListToAnotherBoard;
      const wrapper = shallow(<ListMove store={store} {...props} />);
      const component = wrapper.find('ListMove').dive();

      component.setState({ newBoardId: 'Board2ID', newListPosition: 1 });
      component.find('SubmitButton').simulate('click');

      expect(mockMoveListToAnotherBoard).toHaveBeenCalledTimes(1);
      expect(mockMoveListToAnotherBoard).toHaveBeenCalledWith(expectedAction);
    });
  })
});
