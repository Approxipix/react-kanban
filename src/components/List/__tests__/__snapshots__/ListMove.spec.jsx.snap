// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListMove> should render with given state from Redux store and correct props 1`] = `
<Fragment>
  <styled.div>
    <Select
      label="Board"
      onChange={[Function]}
      options={
        Array [
          Object {
            "title": "Board 1 (current)",
            "value": "Board1ID",
          },
          Object {
            "title": "Board 2",
            "value": "Board2ID",
          },
        ]
      }
      placeholder="Select board"
      value="Board 1"
    />
    <Select
      label="Position"
      onChange={[Function]}
      options={
        Array [
          Object {
            "title": "1 (current)",
            "value": 1,
          },
          Object {
            "title": "2",
            "value": 2,
          },
          Object {
            "title": "3",
            "value": 3,
          },
        ]
      }
      placeholder="Select position"
      value="0"
    />
  </styled.div>
  <SubmitButton
    onClick={[Function]}
  >
    Move
  </SubmitButton>
</Fragment>
`;
